go-controller/pkg/util/util_unit_test.go

```diff
@@ -0,0 +1,242 @@
+			onRetArgsKexecIface:     &onCallReturnArgs{"Command", []string{"string", "string", "string", "string", "string", "string", "string"}, []interface{}{mockCmd}},
+		},
+		{
+			desc:                    "ovs-vsctl command returns empty chassisID along with error",
+			errExpected:             true,
+			onRetArgsExecUtilsIface: &onCallReturnArgs{"RunCmd", []string{"*mocks.Cmd", "string", "[]string", "string", "string", "string", "string", "string", "string"}, []interface{}{bytes.NewBuffer([]byte("")), bytes.NewBuffer([]byte("")), fmt.Errorf("test error")}},
+			onRetArgsKexecIface:     &onCallReturnArgs{"Command", []string{"string", "string", "string", "string", "string", "string", "string"}, []interface{}{mockCmd}},
+		},
+		{
+			desc:                    "ovs-vsctl command returns empty chassisID with NO error",
+			errExpected:             true,
+			onRetArgsExecUtilsIface: &onCallReturnArgs{"RunCmd", []string{"*mocks.Cmd", "string", "[]string", "string", "string", "string", "string", "string", "string"}, []interface{}{bytes.NewBuffer([]byte("")), bytes.NewBuffer([]byte("")), nil}},
+			onRetArgsKexecIface:     &onCallReturnArgs{"Command", []string{"string", "string", "string", "string", "string", "string", "string"}, []interface{}{mockCmd}},
+		},
+		{
+			desc:                    "ovs-vsctl command returns valid chassisID",
+			errExpected:             false,
+			onRetArgsExecUtilsIface: &onCallReturnArgs{"RunCmd", []string{"*mocks.Cmd", "string", "[]string", "string", "string", "string", "string", "string", "string"}, []interface{}{bytes.NewBuffer([]byte("4e98c281-f12b-4601-ab5a-a3d759fcb493")), bytes.NewBuffer([]byte("")), nil}},
+			onRetArgsKexecIface:     &onCallReturnArgs{"Command", []string{"string", "string", "string", "string", "string", "string", "string"}, []interface{}{mockCmd}},
+		},
+	}
+
+	for i, tc := range tests {
+		t.Run(fmt.Sprintf("%d:%s", i, tc.desc), func(t *testing.T) {
+
+			call := mockExecRunner.On(tc.onRetArgsExecUtilsIface.onCallMethodName)
+			for _, arg := range tc.onRetArgsExecUtilsIface.onCallMethodArgType {
+				call.Arguments = append(call.Arguments, mock.AnythingOfType(arg))
+			}
+			for _, ret := range tc.onRetArgsExecUtilsIface.retArgList {
+				call.ReturnArguments = append(call.ReturnArguments, ret)
+			}
+			call.Once()
+
+			ifaceCall := mockKexecIface.On(tc.onRetArgsKexecIface.onCallMethodName)
+			for _, arg := range tc.onRetArgsKexecIface.onCallMethodArgType {
+				ifaceCall.Arguments = append(ifaceCall.Arguments, mock.AnythingOfType(arg))
+			}
+			for _, ret := range tc.onRetArgsKexecIface.retArgList {
+				ifaceCall.ReturnArguments = append(ifaceCall.ReturnArguments, ret)
+			}
+			ifaceCall.Once()
+
+			ret, e := GetNodeChassisID()
+			if tc.errExpected {
+				assert.Error(t, e)
+			} else {
+				assert.Greater(t, len(ret), 0)
+			}
+			mockExecRunner.AssertExpectations(t)
+			mockCmd.AssertExpectations(t)
+		})
```

Comment:
L109-110: Do we really need this calls? Since we are mocking the corresponding methods with expected function arguments  and the return values, they will always be 'True'? So, why assert? L104-107 is what is testing GetNodeChassiID(), right?

Reply:
The AssertExpectations ensures that whatever is specified with `On` and `Return`for the mocks are called as expected. I remember a case in one of the earlier PRs where I specified something incorrectly in the "On"/"Return" and the test would still pass. Having the AssertExpectations helped catch that and make the test more accurate/readable and avoid cut n paste errors for newer tests. For e.g; try modifying L188 from`nil` to `fmt.Errorf("")` (in a real life scenario, say a developer did copy paste and forgot to change from fmt.Errorf to nil ) and run the test case to verify that it fails. Next, if you comment out lines 238 and 239, the test will pass. Someone new that read the test case may end up getting confused and would like to avoid that if at all possible.

Reply:
Thanks for the explaination.